internal/servicelog/ringbuffer.go

```diff
@@ -0,0 +1,369 @@
+	if rb.writeClosed && start+RingPos(written) == rb.writeIndex {
+		return written, io.EOF
+	}
+	return written, nil
+}
+
+// TailIterator returns an iterator from the tail of the buffer.
+func (rb *RingBuffer) TailIterator() Iterator {
+	rb.iteratorMutex.Lock()
+	defer rb.iteratorMutex.Unlock()
+	start, _ := rb.Positions()
+	iter := &iterator{
+		rb:         rb,
+		index:      start,
+		notifyChan: make(chan struct{}),
+		closeChan:  make(chan struct{}),
+	}
+	if rb.Closed() {
+		close(iter.closeChan)
+	}
+	rb.iteratorList = append(rb.iteratorList, iter)
+	return iter
+}
+
+// HeadIterator returns an iterator from the head of the buffer.
+func (rb *RingBuffer) HeadIterator(lines int) Iterator {
+	firstLine := rb.reverseLinePosition(lines)
+	rb.iteratorMutex.Lock()
+	defer rb.iteratorMutex.Unlock()
+	iter := &iterator{
+		rb:         rb,
+		index:      firstLine,
+		notifyChan: make(chan struct{}),
+		closeChan:  make(chan struct{}),
+	}
+	if rb.Closed() {
+		close(iter.closeChan)
+	}
+	rb.iteratorList = append(rb.iteratorList, iter)
+	return iter
+}
+
+func (rb *RingBuffer) reverseLinePosition(n int) RingPos {
+	rb.rwlock.RLock()
+	defer rb.rwlock.RUnlock()
+	if n <= 0 {
+		return rb.writeIndex
+	}
+	buffers := rb.buffers(rb.readIndex, rb.writeIndex)
+	// a line is not complete until newline is written, so start negative.
+	lines := -1
+	firstLine := rb.writeIndex
+	last := byte(0)
+out:
+	for j := len(buffers) - 1; j >= 0; j-- {
+		buf := buffers[j]
+		for i := len(buf) - 1; i >= 0; i-- {
+			firstLine--
+			last = buf[i]
+			if last == byte('\n') {
+				lines++
+			}
+			if lines == n {
+				break out
+			}
+		}
+	}
+	if last == byte('\n') {
+		firstLine++
+	}
+	return firstLine
+}
+
+// Discard disposes of n bytes from the tail of the buffer making
+// them available to be used for subsequent writes.
+func (rb *RingBuffer) Discard(n int) error {
+	rb.rwlock.Lock()
+	defer rb.rwlock.Unlock()
+	return rb.discard(n)
+}
+
+func (rb *RingBuffer) discard(n int) error {
+	buffered := rb.buffered()
+	if n > buffered {
+		n = buffered
+	}
+	rb.readIndex = rb.readIndex + RingPos(n)
+	return nil
+}
+
+func (rb *RingBuffer) signalIterators() {
+	rb.iteratorMutex.RLock()
+	defer rb.iteratorMutex.RUnlock()
+	for _, iter := range rb.iteratorList {
+		select {
+		case iter.notifyChan <- struct{}{}:
+		default:
+		}
+	}
+}
+
+func (rb *RingBuffer) releaseIterators() {
+	rb.iteratorMutex.Lock()
+	defer rb.iteratorMutex.Unlock()
+	for _, iter := range rb.iteratorList {
+		close(iter.closeChan)
+	}
+	rb.iteratorList = nil
+}
+
+func (rb *RingBuffer) removeIterator(iter *iterator) {
+	rb.iteratorMutex.Lock()
+	defer rb.iteratorMutex.Unlock()
+	for i, storedIter := range rb.iteratorList {
+		if iter != storedIter {
+			continue
+		}
+		close(iter.closeChan)
+		rb.iteratorList[i] = rb.iteratorList[len(rb.iteratorList)-1]
+		rb.iteratorList = rb.iteratorList[:len(rb.iteratorList)-1]
+		return
+	}
+}
+
+func (rb *RingBuffer) buffers(start, end RingPos) [2][]byte {
```

Comment:
Would be nice to have a few words documenting this function, as it's not typical nor obvious.